.github/workflows/r_package.yml

```diff
@@ -181,10 +181,33 @@ jobs:
           Rscriptdevel testthat.R 2>&1 > ubsan-tests.log
           cat ubsan-tests.log
           exit $(cat ubsan-tests.log | grep --count "runtime error")
+  test-r-debian-clang:
+    name: r-package (debian, R-devel, clang)
+    timeout-minutes: 60
+    runs-on: ubuntu-latest
+    container: rhub/debian-clang-devel
+    steps:
+      - name: Install Git before checkout
+        shell: bash
+        run: |
+          apt-get update
+          apt-get install --no-install-recommends -y git
+      - name: Checkout repository
+        uses: actions/checkout@v2.3.4
+        with:
+          fetch-depth: 5
+          submodules: true
+      - name: Install packages and run tests
+        shell: bash
+        run: |
+          export PATH=/opt/R-devel/bin/:${PATH}
+          Rscript -e "install.packages(c('R6', 'data.table', 'jsonlite', 'testthat'), repos = 'https://cran.r-project.org')"
+          sh build-cran-package.sh
+          R CMD check --as-cran --run-donttest lightgbm_*.tar.gz || exit -1
```

Comment:
Why don't we check for `ERRORS/WARNINGS/NOTES` here?
https://github.com/microsoft/LightGBM/blob/55a31bfe1a84f26bc65a4c6af3717155f0a6c8d7/.ci/test_r_package.sh#L184-L187

Especially, this looks critical given #3616.

BTW, the same question for `GCC ASAN/UBSAN` above.

Reply:
**short answer**

it's an oversight, and we definitely should! I'll push that change here

**long answer**

All of that indirection where we decide whether or not to fail builds based on grepping through the R CMD check logs is primarily done so we have the chance to echo all of the more detailed logs out first: https://github.com/microsoft/LightGBM/blob/55a31bfe1a84f26bc65a4c6af3717155f0a6c8d7/.ci/test_r_package.sh#L178

When you run `R CMD check --as-cran`, any ERROR will result in a non-0 exit code.
 https://github.com/microsoft/LightGBM/pull/4164#issuecomment-814232054 demonstrates that this worked for a build with an ERROR. 

You're right that these things would miss WARNINGs or NOTEs on those other builds. I'll push a fix shortly.

Reply:
Oh sorry and one other thing...for the UBSAN tests we're just running the unit tests directly with `Rscript testthat.R`, not running all of `R CMD check`.  So that test doesn't need to care about other R CMD check stuff.

Reply:
I just pushed https://github.com/microsoft/LightGBM/pull/4164/commits/1ff93a6fce40856406d7f60893a0b514d09f6e9f, which adds code to fail the debian-clang build on WARNINGs or NOTEs, and intentionally added a documentation mistake that will cause a WARNING, so we can see that work

Reply:
From the descriptions of CRAN's "extra" checks at https://cran.r-project.org/web/checks/check_issue_kinds.html, it looks like the UBSAN/ASAN tests run `R CMD check --use-valgrind`. So it would probably good for us to do that too.

But the feedback cycle when testing a change like that is very slow because R runs so much more slowly when instrumented with valgrind (those jobs often take around an hour to run). I'll write that up in a separate feature request. It isn't critical for this PR or the 3.2.1 release.

Reply:
Ok I'm pretty confident the new change will fail builds on warnings: https://github.com/microsoft/LightGBM/pull/4164/checks?check_run_id=2313684939

```text
Status: 1 WARNING
See
  '/__w/LightGBM/LightGBM/lightgbm.Rcheck/00check.log'
for details.


NOTEs, WARNINGs, or ERRORs have been found by R CMD check
Error: Process completed with exit code 255.
```

I'll push one change that will throw a NOTE to test that it also fails on NOTEs.

Reply:
Here is a build failing on a NOTE:
https://github.com/microsoft/LightGBM/pull/4164/checks?check_run_id=2313911147

```text
Status: 1 NOTE
See
  '/__w/LightGBM/LightGBM/lightgbm.Rcheck/00check.log'
for details.


NOTEs, WARNINGs, or ERRORs have been found by R CMD check
Error: Process completed with exit code 255.
```

Reply:
Thank you very much for doing this!